[optional.ctor], [expected.object.ctor] Add converts-from-any-cvref (#5416)
cplusplus/draft: source/utilities.tex

\rSec3[optional.ctor]{Constructors}

\pnum
The exposition-only alias template \exposid{converts-from-any-cvref}
is used by some constructors for \tcode{optional}.
\begin{codeblock}
template<class T, class W>

> using @\exposid{converts-from-any-cvref}@ // \expos

Reviewer: ```suggestion
using @\exposid{converts-from-any-cvref}@  // \expos
```
> [If possible, there should be at least two spaces between code and comment.](https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines)

Isn't this supposed to be enforced by a check?

> inline constexpr bool @\exposid{converts-from-any-cvref}@  // \expos

Reviewer 2: We don't need "inline" for variable templates; please drop.

Reviewer 2: See #4625

Author: Done

> using @\exposid{converts-from-any-cvref}@  // \expos
>   = disjunction<is_constructible<T, W&>, is_constructible<T, W>,

Reviewer 3: This should be a variable template or concept instead of an alias template (1) because it's easier to use that way, and (2) because you're already using it as if it is.
```suggestion
inline constexpr bool @\exposid{converts-from-any-cvref}@  // \expos
  = disjunction_v<is_constructible<T, W&>, is_constructible<T, W>,
```

Author: Oops, yes, it needed `::value` on the uses, thanks for catching that.

>   = disjunction_v<is_constructible<T, W&>, is_constructible<T, W>,

Reviewer 2: put "=" on previous line

Author: Done

>                 is_constructible<T, const W&>, is_constructible<T, const W>,
>                 is_convertible<W&, T>, is_convertible<W, T>,
>                 is_convertible<const W&, T>, is_convertible<const W, T>>;

Reviewer: ```suggestion
                 is_constructible<T, const W&>, is_constructible<T, const W>,
                 is_convertible<W&, T>, is_convertible<W, T>,
                 is_convertible<const W&, T>, is_convertible<const W, T>>;
```

>                  is_convertible<const W&, T>, is_convertible<const W, T>>;

Reviewer 2: We're spending four lines here. Shouldn't we group the is_constructibe and is_convertible on a single line (for the same pair of arguments, just reversed because reasons)?

Author: Done

\begin{itemize}
\item \tcode{is_constructible_v<T, U>} is \tcode{true},
\item \tcode{is_assignable_v<T\&, U>} is \tcode{true},

> \item \tcode{\exposid{converts-from-any-cvref}<T, optional<U>>} is \tcode{false}.

Reviewer: ```suggestion
\item \tcode{\exposid{converts-from-any-cvref}<T, optional<U>>} is \tcode{false},
```
